internal/addrs/partial_expanded.go

```diff
@@ -258,6 +413,99 @@ type PartialExpandedResource struct {
 	resource Resource
 }
 
+// ParsePartialExpandedResource parses a resource address traversal and returns
+// a PartialExpandedResource representing the known and unknown parts of the
+// address.
+func ParsePartialExpandedResource(traversal hcl.Traversal) (PartialExpandedResource, hcl.Traversal, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+
+	pem, remain, diags := ParsePartialExpandedModule(traversal)
+	if len(remain) == 0 {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid address",
+			Detail:   "Resource address must be a module address followed by a resource address.",
+			Subject:  traversal.SourceRange().Ptr(),
+		})
+		return PartialExpandedResource{}, nil, diags
+	}
+
+	mode := ManagedResourceMode
+	if remain.RootName() == "data" {
+		mode = DataResourceMode
+		remain = remain[1:]
+	}
+
+	if len(remain) < 2 {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid address",
+			Detail:   "Resource specification must include a resource type and name.",
+			Subject:  remain.SourceRange().Ptr(),
+		})
+		return PartialExpandedResource{}, nil, diags
+	}
+
+	var typeName, name string
+	switch tt := remain[0].(type) {
+	case hcl.TraverseRoot:
+		typeName = tt.Name
+	case hcl.TraverseAttr:
```

Comment:
I don't think we could have a `TraverseAttr` here, since there is only one resource address following a module, and we turned that into a `TraverseRoot`

Reply:
As in the other comment, for a data source this would be `TraverseAttr` as we consumed the `TraverseRoot` when processing the `data` entry.